ext/standard/pack.c

```diff
@@ -870,17 +870,31 @@ PHP_FUNCTION(unpack)
 			RETURN_FALSE;
 		}
 
+
 		/* Do actual unpacking */
-		for (i = 0; i != arg; i++ ) {
-			/* Space for name + number, safe as namelen is ensured <= 200 */
-			char n[256];
+		for (i = 0; i != repetitions; i++ ) {
+			zend_string* real_name;
+			zval val;
+			char buffer[32];
+
+			if (repetitions == 1 && namelen > 0) {
+				/* Use a part of the formatarg argument directly as the name. */
+				real_name = zend_string_init_fast(name, namelen);
 
-			if (arg != 1 || namelen == 0) {
-				/* Need to add element number to name */
-				snprintf(n, sizeof(n), "%.*s%d", namelen, name, i + 1);
 			} else {
-				/* Truncate name to next format code or end of string */
-				snprintf(n, sizeof(n), "%.*s", namelen, name);
+				/* Need to add the 1-based element number to the name */
+
+				/* hardcoded case for small-ish number of repetitions */
+				if ((i+1) < 100) {
+					buffer[0] = '0' + (i+1) / 10;
+					buffer[1] = '0' + (i+1) % 10;
+					int digits = (i+1) < 10 ? 1 : 2;
+					real_name = zend_string_concat2(name, namelen, buffer + 2 - digits, digits);
+
+				} else {
+					int digits = snprintf(buffer, sizeof(buffer), "%d", i+1);
+					real_name = zend_string_concat2(name, namelen, buffer, digits);
+				}
```

Comment:
You could use zend_print_ulong_to_buf. The API is slightly arcane, you can take a look at zend_ulong_to_str on how to use it. zend_print_ulong_to_buf + zend_string_concat2 should give reasonable performance.

Reply:
this is exactly it, thaks